Improve parameterize docs (#18873)
pulumi/pulumi: docs/architecture/providers/parameterized.md

(parameterized-providers)=

> # Parameterized providers

Reviewer: ```suggestion
# Parameterized Providers
```

*Parameterized providers* are a provider feature where a user can generate and

> use a custom SDK based on some custom input.

Reviewer 2: This isn't strictly true I don't think -- the set of use cases is a bit wider than this (though of course this is a common user-visible artifact of it) and I do think the original wording is closer to the truth here. That's not to say we shouldn't talk about custom SDKs, etc., but that's more around `pulumi package add`/the consequences of parameterization, if that makes sense, rather than parameterization itself.

> Key rules:

Reviewer: ```suggestion
Requirements and Conditions:
```

> * Each provider parameterization is run in its own provider instance.

Reviewer 2: This is true right now, for replacement parameterization, but I don't think (?) it will be for extensions. There, you might parameterize an instance multiple times. Not to say we shouldn't talk about this of course, just that we could contextualise to replacement.

Parameterized provider instance. The version of the parameterized package can be
anything the provider would like and does not have to align to the version of the
base provider. The version does not have any specific purpose right now - it's
purely informational, but is required to be set.

Once [](pulumirpc.ResourceProvider.Parameterize) has been called (in either mode)
the engine must be able to call GetSchema and Configure to start to use the
provider or generate the SDK. For replacement parameterization, once Parameterize
has been called, the provider will only be used in its parameterized form.

Once parameterized, GetSchema can be called by the engine. The GetSchema request
will contain the same name and version information as was returned from
[](pulumirpc.ResourceProvider.Parameterize) in the SubpackageName and
SubpackageVersion fields, respectively. The generated schema *must* have the same
name and version as the SubpackageName and SubpackageVersion. GetSchema must be
callable after either the "args" or "value" mode of parameterize was called. It's
expected that GetSchema doesn't have to be fast and that the engine will cache

> schemas as required.

Reviewer 2: I like this content but I wonder about intermingling the "reference" documentation (what parameterization is, how it works) with "tutorial"-esque documentation (how you should implement it, what you can expect) is muddying some of the water here. I feel like potentially breaking out this kind of content into a section, or putting it in the existing implementer's guide, might be a better fit?

Author: I've separated this a bit more within this document. Can always add it somewhere else later and remove from here.

  base-64-encoded when embedded in the SDK.

(provider-implementation)=
## Provider implementation

Requirements and conditions for "replacement" parameterization:

* Each provider parameterization is run in its own provider instance.
* Parameterized package names must be unique within the program where its added.
* Parameterization is either via CLI args *or* embedded metadata.
* [](pulumirpc.ResourceProvider.Parameterize) is always called before [](pulumirpc.ResourceProvider.Configure).
* [](pulumirpc.ResourceProvider.Parameterize) is always called before [](pulumirpc.ResourceProvider.GetSchema).

:::{note}
Currently, we only support "replacement" parameterization where generated SDKs
don't reference the provider's original SDK. See
[Replacement and extension parameterization](#replacement-extension-parameterization)
for discussion of what alternative implementations might be possible in the future.
:::

For a provider to support parameterization it must implement the
[](pulumirpc.ResourceProvider.Parameterize) call. The
[](pulumirpc.ParameterizeRequest) will either contain the args passed via the
CLI from `pulumi package add` or `pulumi package gen-sdk`; or will be passed the

> previous generated metadata binary blob. [](pulumirpc.ResourceProvider.Parameterize)

Reviewer 3: ```suggestion
previous generated metadata binary blob. For any fixed set of parameters, the "args" call will come first and only once; every subsequent call will be of the "blob" form. [](pulumirpc.ResourceProvider.Parameterize)
```

  a parameter describing a set of CRDs, enabling it to then extend its schema to
  expose them to programs and SDK generation.

As hinted at by the above examples, [](pulumirpc.ResourceProvider.Parameterize)
encodes a provider-specific *parameter* that is used to influence the provider's
behaviour. The parameter passed in the [](pulumirpc.ParameterizeRequest) can
take two forms, corresponding to the two contexts in which parameterization
typically occurs:

* When generating an SDK (e.g. using a `pulumi package add` command), we need to
  boot up a provider and parameterize it using only information from the
  command-line invocation. In this case, the parameter is a string array
  representing the command-line arguments (`args`).
* When interacting with a provider as part of program execution, the parameter

>   is *embedded in the SDK*, so as to free the program author from having to know

Reviewer 3: Agreed, this whole paragraph is useful.

Although it's missing an example: what would a provider author want to persist in this way?

>   is *embedded in the SDK*, so as to free the program author from having to know

Author: I feel like the whole terminology is kinda off w.r.t. providers. What's refered to as the "binary parameter" is what we'd always call the "provider metadata" - the internal data required for the provider to fuction correctly.

> resource's package name matches exactly the name of the provider

Reviewer: ```suggestion
resource's package name exactly matches the name of the provider
```

(replacement-extension-providers)=

> ## Replacement and extension parameterization

Reviewer: ```suggestion
## Replacement and Extension Parameterization
```

Reviewer 2: I think generally we've opted for sentence case in most of our developer docs so while it's minor I'd probably not do this for consistency. But maybe just me

Currently, the only kind of parameterization is *replacement parameterization*.

> In this mode, once the provider has been parameterized, it will only be used for

Reviewer: ```suggestion
In this mode, once the provider has been parameterized, it will be used for
```

> that parameter only; and not without the parameter or with another parameter.

Reviewer 2: Thanks for adding this. Such a simple thing we should have written down in the first place!

In the future we might creation the option for a parameterization to be an
"extension" of the original schema. Details are yet to be worked out here, but we
expect the following differences from the replacement parameterization:

> * The provider for the extension SDK might have to be the same as the base provider

Reviewer 2: I think it'd be good to understand what we want to achieve by hypothesising about the design in these docs -- feels like maybe we're at risk of writing stuff that may turn out to be false/not useful later on?

> * The provider for the extension SDK might have to be the same as the base provider

Reviewer 3: Let's just not talk about potential future stuff in the docs, except as a very high-level summary with a pointer to a GH issue.

> * The provider for the extension SDK might have to be the same as the base provider

Author: The only reason to include it here is because it's already being discussed so it's useful to explain why we distinguish the current implementation as "replacement"

> * Tokens in the extension schema must have the same provider name as the base

Author: I've removed these details.

for example).
:::

(replacement-extension-parameterization)=
## Replacement and extension parameterization

> Currently, the only kind of parameterization is *replacement parameterization*.
> In this mode, once the provider has been parameterized, it will be used for that
> parameter only; and not without the parameter or with another parameter.

Reviewer 3: ```suggestion
Currently, the only kind of parameterization is *replacement parameterization*. 
In this mode, the parameterization of the provider _replaces_ the schema and any additional data the provider uses. Once the provider has been parameterized, it will be used for those parameters only; and not without the parameters or with other parameters.
```

The replacement of the schema, instead of managing additions, was a stumbling block for me initially.
